File: problems/29-advanced-trie/003-prefix-frequency-deletion/java/Solution.java
import java.util.*;

public class Solution {
    static class Node { Node[] next = new Node[26]; int pass, end; }
    private final Node root = new Node();

    public void insert(String word) {
        Node cur = root; cur.pass++;
        for (char ch : word.toCharArray()) {
            int i = ch - 'a';
            if (cur.next[i] == null) cur.next[i] = new Node();
            cur = cur.next[i]; cur.pass++;
        }
        cur.end++;
    }

    public boolean delete(String word) {
        Node[] path = new Node[word.length() + 1];
        Node cur = root; path[0] = root;
        for (int p = 0; p < word.length(); p++) {
            int i = word.charAt(p) - 'a';
            if (cur.next[i] == null || cur.next[i].pass == 0) return false;
            cur = cur.next[i]; path[p + 1] = cur;
        }
        if (cur.end == 0) return false;
        cur.end--;
        for (Node node : path) node.pass--;
        return true;
    }

    public int prefixCount(String prefix) {
        Node cur = root;
        for (char ch : prefix.toCharArray()) {
            cur = cur.next[ch - 'a'];
            if (cur == null) return 0;
        }
        return cur.pass;
    }
}

File: problems/29-advanced-trie/003-prefix-frequency-deletion/python/solution.py
class Node:
    def __init__(self):
        self.next = {}
        self.pass_count = 0
        self.end = 0

class Trie:
    def __init__(self):
        self.root = Node()

    def insert(self, word):
        cur = self.root
        cur.pass_count += 1
        for ch in word:
            cur = cur.next.setdefault(ch, Node())
            cur.pass_count += 1
        cur.end += 1

    def delete(self, word):
        cur = self.root
        path = [cur]
        for ch in word:
            if ch not in cur.next or cur.next[ch].pass_count == 0:
                return False
            cur = cur.next[ch]
            path.append(cur)
        if cur.end == 0:
            return False
        cur.end -= 1
        for node in path:
            node.pass_count -= 1
        return True

    def prefix_count(self, prefix):
        cur = self.root
        for ch in prefix:
            if ch not in cur.next:
                return 0
            cur = cur.next[ch]
        return cur.pass_count
